cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,415 @@
+   *
+   * @note `lhs` and `rhs` may be the same.
+   *
+   * @param has_nulls Indicates if either input column contains nulls.
+   * @param lhs The column containing the first element
+   * @param rhs The column containing the second element (may be the same as lhs)
+   * @param null_precedence Indicates how null values are ordered with other values
+   * @param depth The depth of the column if part of a nested column @see preprocessed_table::depths
+   */
+  __device__ element_comparator(Nullate has_nulls,
+                                column_device_view lhs,
+                                column_device_view rhs,
+                                null_order null_precedence = null_order::BEFORE,
+                                int depth                  = 0)
+    : _lhs{lhs}, _rhs{rhs}, _nulls{has_nulls}, _null_precedence{null_precedence}, _depth{depth}
+  {
+  }
+
+  /**
+   * @brief Performs a relational comparison between the specified elements
+   *
+   * @param lhs_element_index The index of the first element
+   * @param rhs_element_index The index of the second element
+   * @return Indicates the relationship between the elements in the `lhs` and `rhs` columns, along
+   * with the depth at which a null value was encountered.
+   */
+  template <typename Element, CUDF_ENABLE_IF(cudf::is_relationally_comparable<Element, Element>())>
+  __device__ cuda::std::pair<weak_ordering, int> operator()(
+    size_type const lhs_element_index, size_type const rhs_element_index) const noexcept
+  {
+    if (_nulls) {
+      bool const lhs_is_null{_lhs.is_null(lhs_element_index)};
+      bool const rhs_is_null{_rhs.is_null(rhs_element_index)};
+
+      if (lhs_is_null or rhs_is_null) {  // at least one is null
+        return cuda::std::make_pair(null_compare(lhs_is_null, rhs_is_null, _null_precedence),
+                                    _depth);
+      }
+    }
+
+    return cuda::std::make_pair(relational_compare(_lhs.element<Element>(lhs_element_index),
+                                                   _rhs.element<Element>(rhs_element_index)),
+                                std::numeric_limits<int>::max());
+  }
+
+  template <typename Element,
+            CUDF_ENABLE_IF(not cudf::is_relationally_comparable<Element, Element>() and
+                           not std::is_same_v<Element, cudf::struct_view>)>
+  __device__ cuda::std::pair<weak_ordering, int> operator()(size_type const lhs_element_index,
+                                                            size_type const rhs_element_index)
+  {
+    // TODO: make this CUDF_UNREACHABLE
+    cudf_assert(false && "Attempted to compare elements of uncomparable types.");
+    return cuda::std::make_pair(weak_ordering::LESS, std::numeric_limits<int>::max());
+  }
+
+  template <typename Element,
+            CUDF_ENABLE_IF(not cudf::is_relationally_comparable<Element, Element>() and
+                           std::is_same_v<Element, cudf::struct_view>)>
```

Comment:
https://godbolt.org/z/3vf6Y7oeq

Due to an annoying gcc bug, you have to put explicit specializations outside of the type.

Reply:
Same problem as https://github.com/rapidsai/cudf/pull/10164#discussion_r820461944 but I should remove that other check from the enable_if

Reply:
Oh, I missed the fact that we're inside a class template. Okay, yeah, can't easily be done.